Fixed REPL for `:doc Type` to display documentation rather than giving an error

RossMeikleham opened the pull request:
Fix for #2273 

REPL now displays
```
Idris> :doc Type
Type : Type 1
    The type of types
```

rather than
```
Idris> :doc Type
Usage is :doc <functionname>
```

david-christiansen commented on src/Idris/REPL.hs:
```diff
@@ -942,19 +942,21 @@ process fn (Core t)
-          modDocN (NS (UN n) ns) = NS modDocName (n:ns)
-          modDocN (UN n)         = NS modDocName [n]
-          modDocN _              = sMN 1 "NotFoundForSure"
+        case show n of
```
Shouldn't this pattern-match the name itself and check for the specific name that you created, rather than relying on the Show instance not changing? That would be less prone to breakage if we e.g. add universe level display.

david-christiansen commented:
Thanks!  I just saw that you had posted a request for help on #2273 that I had missed - sorry! It's OK to ping from time to time with these things, or email the mailing list if people don't see the Github activity.

This implementation looks quite reasonable. I'm a bit concerned about the pattern match on Show, however, - is there a reason not to analyze the structure of the name itself?

RossMeikleham commented:
No worries, I'll make sure to do that next time.

Yup it would be better to pattern match on the name, would the following be more appropriate?
```Idris
   = do ist <- getIState
        let txtType = txt "Type" 
        case n of
          (UN txtType) -> iRenderResult $ pprintType1Doc ist
          _ -> ...
```

david-christiansen commented:
Yeah, that looks nicer.  Perhaps even better would be to replace the Either with a "DocArg" datatype with constructors for constants, names, Type, etc, but this can perhaps wait until one more thing gets docs in Idris.

RossMeikleham commented:
Ok, just updated my PR with the pattern matching changes and renamed one of the functions I introduced in the last commit `pprintType1Doc` to `pprintTypeDoc`.

david-christiansen commented:
Thanks again! Travis is doing its thing.

Looking for a new project?

RossMeikleham commented:
No problem. I'm starting an internship next week so I won't have much free time as I'd like, but I'd be happy to work on a project in some of my spare time if you have any suggestions.

david-christiansen commented:
If you like working with the parts of the compiler that you've already seen, #2241 and #2176 would make sense.

RossMeikleham commented:
Just realised I've made a really silly mistake in my last PR which if it hasn't been picked up by the travis it soon will.  The pattern matched `txtType` is just matching any value with the UN constructor causing lots of docs to become the same as Type. Trying to find another way around it but the UN constructor takes a "Text" type from `Data.Text`, and there doesn't seem to be any explicit constructor for it in that module.

An explicit equality check works, but that relies on the `Eq` instance for `Name` staying the
same:

```Idris
        if n == (UN $ txt "Type")
          then iRenderResult $ pprintTypeDoc ist
          else do let docs = lookupCtxtName n (idris_docstrings ist) ++
```

Unless there is a constructor for `Text` types I'm missing.

RossMeikleham commented:
Another possible solution is the following although not sure if it's too "messy":

```Idris
process fn (DocStr (Left n) w)
   = do ist <- getIState
        let txtType = txt "Type" 
        case n of
          (UN a) -> case (str a) of
                "Type" -> iRenderResult $ pprintTypeDoc ist
                _ -> processNotType ist
          _ -> processNotType ist

        where processNotType ist = do
              let docs = lookupCtxtName n (idris_docstrings ist) ++
                             map (\(n,d)-> (n, (d, [])))
                                 (lookupCtxtName (modDocN n) (idris_moduledocs ist))
              case docs of
                [] -> iPrintError $ "No documentation for " ++ show n
                ns -> do toShow <- mapM (showDoc ist) ns
                         iRenderResult (vsep toShow)
              where showDoc ist (n, d) = do doc <- getDocs n w
                                            return $ pprintDocs ist doc

                    modDocN (NS (UN n) ns) = NS modDocName (n:ns)
                    modDocN (UN n)         = NS modDocName [n]
                    modDocN _              = sMN 1 "NotFoundForSure"

```

david-christiansen commented:
It's fairly common in the Idris codebase to see things like:

````
process fn (DocStr (Left n) w)
  | UN ty <- n, ty == T.pack "Type" = ....
  | otherwise =  ...
````
That idiom seems to work fairly well.

RossMeikleham commented:
Thanks! I didn't know it was possible to perform pattern matching in guards. I've updated my PR (remembered to run the tests on my machine this time)

david-christiansen commented:
Thanks!
